### PR TITLE
Added MirrorOfMatrix.java

### PR DESCRIPTION
**Pull request by Aman28801**
Problem Statement :-
We have given an array of m x n (where m is the number of rows and n is the number of columns).
Print the new matrix in such a way that the new matrix is the mirror image of the original matrix.
Solution;_
it basically traverse column wise and swap the starting and ending element of each row to get the mirror of 2D array

In this pull request, I am submitting the code of this problem that prioritize readability and user-friendliness.

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.

### CODE REVIEW

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        System.out.println("The Original matrix is:");
+        for (int i = 0; i < row; i++) {
+            for (int j = 0; j < column; j++) {
+                System.out.print(arr[i][j] + "\t");
+            }
+            System.out.print("\n");
+        }
+        // Calling Function Mirror
+        Mirror(arr);
+    }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:68]:
```suggestion
```
Please remove the `main` method and instead add some proper tests. Good examples to look at:
- [`Factorial`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/Factorial.java) and [`FactorialTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FactorialTest.java),
- [`FindMax`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/FindMax.java) and [`FindMaxTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FindMaxTest.java).

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+Print the new matrix in such a way that the new matrix is the mirror image of the original matrix.
+
+The Original matrix is:   |   The Mirror matrix is:
+1	2	3	              |   3	 2	 1
+4	5	6	              |   6	 5	 4
+7	8	9	              |   9	 8	 7
+
+@author - Aman (https://github.com/Aman28801)
+*/
+public class MirrorOfMatrix {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:17]: Please make it a _proper utility class_.

```suggestion
public final class MirrorOfMatrix {
    private MirrorOfMatrix() {
    }
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java (hunk 2 above)
Comment (Aman28801) [reply src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:17]: this is done

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java (hunk 1 above)
Comment (Aman28801) [reply src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:68]: in this i am not understanding the difference in both of them

**Comment by Aman28801** [comment]
I would be grateful if my contribution is accepted and merged using "hacktoberfest-accepted" label.

**Review by vil02 — CHANGES_REQUESTED** [review]
As mentioned [earlier](https://github.com/TheAlgorithms/Java/pull/4461#discussion_r1342018901): please add tests for your class.

**Comment by Aman28801** [comment]
is test file okay?

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+4	5	6	              |   6	 5	 4
+7	8	9	              |   9	 8	 7
+
+@author - Aman (https://github.com/Aman28801)
+*/
+
+public final class MirrorOfMatrix {
+    private MirrorOfMatrix() {
+    }
+    // Function Mirror that is used to mirror the Matrix
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:21]:
```suggestion
```

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package com.thealgorithms.misc;
+
+import java.util.Scanner;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:3]:
```suggestion
```

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+7	8	9	              |   9	 8	 7
+
+@author - Aman (https://github.com/Aman28801)
+*/
+
+public final class MirrorOfMatrix {
+    private MirrorOfMatrix() {
+    }
+    // Function Mirror that is used to mirror the Matrix
+    public static void Mirror(final int[][] arr) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:22]: Why do you modify the input and not return a new array?

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                hi--;
+            }
+        }
+        // For printing the output of Code OR Mirror Image of Matrix
+        System.out.println("The Mirror matrix is:");
+        for (int[] ints : arr) {
+            for (int j = 0; j < arr[0].length; j++) {
+                System.out.print(ints[j] + "\t");
+            }
+            System.out.println();
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:42]: This is not needed.
```suggestion
```

File: src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        int[][] arr = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
+        MirrorOfMatrix.Mirror(arr);
+        // The expected result is each row reversed.
+        int[][] expected = {{3, 2, 1}, {6, 5, 4}, {9, 8, 7}};
+        assertArrayEquals(expected, arr);
+    }
+
+    @Test
+    public void testMirrorWithOddLengthRows() {
+        int[][] arr = {{1, 2, 3}, {4, 5, 6, 7}, {8, 9}};
```
Comment (vil02) [inline src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java]: This is not a matrix. What kind of inputs do you allow?

File: src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        MirrorOfMatrix.Mirror(arr);
+        // The expected result is each row reversed.
+        int[][] expected = {{3, 2, 1}, {7, 6, 5, 4}, {9, 8}};
+        assertArrayEquals(expected, arr);
+    }
+}
```
Comment (vil02) [inline src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java]:
```suggestion
}

```

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+We have given an array of m x n (where m is the number of rows and n is the number of columns).
+Print the new matrix in such a way that the new matrix is the mirror image of the original matrix.
+
+The Original matrix is:   |   The Mirror matrix is:
+1	2	3	              |   3	 2	 1
+4	5	6	              |   6	 5	 4
+7	8	9	              |   9	 8	 7
+
+@author - Aman (https://github.com/Aman28801)
+*/
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:16]: Please update the description of the problem. Try to explain in what sentence what your method does.

File: src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        // So, the expected result is also an empty matrix.
+        int[][] expected = new int[0][0];
+        assertArrayEquals(expected, arr);
+    }
+
+    @Test
+    public void testMirrorWithSingleRowMatrix() {
+        int[][] arr = {{1, 2, 3, 4}};
+        MirrorOfMatrix.Mirror(arr);
+        // The expected result is the reverse of the input row.
```
Comment (vil02) [inline src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java]:
```suggestion
```

File: src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        int[][] expected = {{4, 3, 2, 1}};
+        assertArrayEquals(expected, arr);
+    }
+
+    @Test
+    public void testMirrorWithMultipleRowMatrix() {
+        int[][] arr = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
+        MirrorOfMatrix.Mirror(arr);
+        // The expected result is each row reversed.
```
Comment (vil02) [inline src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java]:
```suggestion
```

File: src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        int[][] expected = {{3, 2, 1}, {6, 5, 4}, {9, 8, 7}};
+        assertArrayEquals(expected, arr);
+    }
+
+    @Test
+    public void testMirrorWithOddLengthRows() {
+        int[][] arr = {{1, 2, 3}, {4, 5, 6, 7}, {8, 9}};
+        MirrorOfMatrix.Mirror(arr);
+        // The expected result is each row reversed.
```
Comment (vil02) [inline src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java]:
```suggestion
```

File: src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.jupiter.api.Test;
+
+public class MirrorOfMatrixTest {
+
+    @Test
+    public void testMirrorWithEmptyMatrix() {
+        int[][] arr = new int[0][0];
+        MirrorOfMatrix.Mirror(arr);
+        // Since the matrix is empty, there's nothing to mirror.
+        // So, the expected result is also an empty matrix.
```
Comment (vil02) [inline src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java]:
```suggestion
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java (hunk 7 above)
Comment (Aman28801) [reply src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java]: 2D array inputs

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java (hunk 9 above)
Comment (Aman28801) [reply src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:15]: done

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java (hunk 5 above)
Comment (Aman28801) [reply src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:20]: not take extra space

File: src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java (hunk 7 above)
Comment (vil02) [reply src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java]: So do you actually not allow such inputs? You wrote that you expect a matrix as an input.

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java (hunk 5 above)
Comment (vil02) [reply src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:22]: On the other hand modifying the input is a bad practice.

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java (hunk 9 above)
Comment (vil02) [reply src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:16]: Please write a proper documentation of your class. The method `Mirror` does not print anything.

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+public final class MirrorOfMatrix {
+    private MirrorOfMatrix() {
+    }
+    public static int[][] mirrorMatrix(int[][] originalMatrix) {
+        if (originalMatrix == null || originalMatrix.length == 0) {
+            // Handle invalid input
+            return null;
+        }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:25]: Why if the input is an empty array, the output is `null`? It should be an empty array in my opinion.

**Review by vil02 — CHANGES_REQUESTED** [review]
What have happened with the other tests? Especially with the one with empty input?

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+@author - Aman (https://github.com/Aman28801)
+*/
+
+public final class MirrorOfMatrix {
+    private MirrorOfMatrix() {
+    }
+    public static int[][] mirrorMatrix(int[][] originalMatrix) {
+        if (originalMatrix == null || originalMatrix.length == 0) {
+            // Handle invalid input
+            return new int[0][0];
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:22]: If the `orginalMatrix` is `null` returning `null` makes sense.

File: src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        int[][] singleRowMatrix = {{1, 2, 3}};
+        int[][] expectedSingleRowMirroredMatrix = {{3, 2, 1}};
+        int[][] mirroredSingleRowMatrix = MirrorOfMatrix.mirrorMatrix(singleRowMatrix);
+        assertArrayEquals(expectedSingleRowMirroredMatrix, mirroredSingleRowMatrix);
+
+        // Test case 4: Single-column matrix
+        int[][] singleColumnMatrix = {{1}, {2}, {3}};
+        int[][] expectedSingleColumnMirroredMatrix = {{1}, {2}, {3}};
+        int[][] mirroredSingleColumnMatrix = MirrorOfMatrix.mirrorMatrix(singleColumnMatrix);
+        assertArrayEquals(expectedSingleColumnMirroredMatrix, mirroredSingleColumnMatrix);
```
Comment (vil02) [inline src/test/java/com/thealgorithms/misc/MirrorOfMatrixTest.java:34]: Please split it into separate methods or (even better!) use _parametrized tests_ as in [FindMaxTest](https://github.com/TheAlgorithms/Java/blob/master/src/test/java/com/thealgorithms/maths/FindMaxTest.java).

And please add test cases like _empty input_, `null` input, _input with empty rows_.

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+        int numRows = originalMatrix.length;
+        int numCols = originalMatrix[0].length;
+
+        int[][] mirroredMatrix = new int[numRows][numCols];
+
+        for (int i = 0; i < numRows; i++) {
+            for (int j = 0; j < numCols; j++) {
+                // Copy elements from the original matrix in reverse order
+                mirroredMatrix[i][j] = originalMatrix[i][numCols - 1 - j];
+            }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:33]:
```suggestion
            mirroredMatrix[i] = reverseRow(originalMatrix[i]);
```
with:
```java
    private static int[] reverseRow(final int[] inRow) {
        int[] res = new int[inRow.length];
        for (int i = 0; i < inRow.length; ++i) {
            res[i] = inRow[inRow.length - 1 - i];
        }
        return res;
    }
```
The advantage is that if someone reads `mirroredMatrix[iRow] = reverseRow(originalMatrix[iRow]);` it is clear from the method name what happens there.

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+public final class MirrorOfMatrix {
+    private MirrorOfMatrix() {
+    }
+    public static int[][] mirrorMatrix(int[][] originalMatrix) {
+        if (originalMatrix == null || originalMatrix.length == 0) {
+            // Handle invalid input
+            return new int[0][0];
+        }
+        int numRows = originalMatrix.length;
+        int numCols = originalMatrix[0].length;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:25]: So what will happen in case of an input  not being a real matrix (i.e. having _rows_ with different sizes)?

If you want to exclude such possibility (your current implementation _assumes_ that the input is a proper matrix) I would suggest to throw `IllegalArgumentException`.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+The Original matrix is:   |   The Mirror matrix is:
+1	2	3	              |   3	 2	 1
+4	5	6	              |   6	 5	 4
+7	8	9	              |   9	 8	 7
+
+@author - Aman (https://github.com/Aman28801)
+*/
+
+public final class MirrorOfMatrix {
+    MirrorOfMatrix() {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:17]:
```suggestion
    private MirrorOfMatrix() {
```

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+            return new int[0][0];
+        }
+        int numRows = originalMatrix.length;
+        int numCols = originalMatrix[0].length;
+
+        for (int i = 1; i < numRows; i++) {
+            if (originalMatrix[i].length != numCols) {
+                throw new IllegalArgumentException("Input matrix is not a valid matrix. Rows have different sizes.");
+            }
+        }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:34]: What do you think about moving this (i.e. _abstracting_) into a separate method like `checkInput(...)`.

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+        if (originalMatrix.length == 0) {
+            return new int[0][0];
+        }
+        int numRows = originalMatrix.length;
+        int numCols = originalMatrix[0].length;
+
+        for (int i = 1; i < numRows; i++) {
+            if (originalMatrix[i].length != numCols) {
+                throw new IllegalArgumentException("Input matrix is not a valid matrix. Rows have different sizes.");
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:32]:
```suggestion
                throw new IllegalArgumentException("The input is not a matrix.");
```

By the way: this line is not covered by any of the tests. Please have a look here:
https://github.com/TheAlgorithms/Java/blob/536978919dc79af9bf25c57ebb90802d1830784e/src/test/java/com/thealgorithms/maths/NthUglyNumberTest.java#L72

File: src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+            return null;
+        }
+        if (originalMatrix.length == 0) {
+            return new int[0][0];
+        }
+        int numRows = originalMatrix.length;
+        int numCols = originalMatrix[0].length;
+
+        for (int i = 1; i < numRows; i++) {
+            if (originalMatrix[i].length != numCols) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/misc/MirrorOfMatrix.java:31]:
```suggestion
            if (originalMatrix[i].length != originalMatrix[0].length) {
```

**Review by vil02 — CHANGES_REQUESTED** [review]

**Comment by Aman28801** [comment]
I would be grateful if my contribution is accepted and merged using "hacktoberfest-accepted" label.

**Comment by vil02** [comment]
> I would be grateful if my contribution is accepted and merged using "hacktoberfest-accepted" label.

It is not needed: the hacktoberfest page will recognize this contribution, because this repo has hactoberest as a one of the topics.

@Aman28801 please have a look at this commits: ec8e9f820129f8ddf2af6022c8fda7bdb77f2931

**Review by vil02 — APPROVED** [review]

**Comment by Aman28801** [comment]
Thank you for the help @vil02 throughout the pull request as this is my first contribution towards open source.
